### DISCUSSION TITLE
Is there a default delay applied to resonses?

### DISCUSSION THREAD

**Discussion by philals** [mswjs/msw#553] [Q&A] [answered]
Hi - thanks for making MSW. It's grand.

We have been noticing a bit of lag for MSW to respond to requests. We have never used ctx.delay()

Reading the docs on delay() it was not clear: Does MSW apply a default random delay to responses? Even when never calling delay()?

Thanks

**Comment by kettanaito** [comment, marked as answer]
Hi, @philals. Thanks for reaching out. 

MSW _does not_ apply any default delay to mocked responses. You can see the default response properties here: 

https://github.com/mswjs/msw/blob/b99be78eede52ebde08c48fc78f773cce3840a6c/src/response.ts#L38

The random realistic response delay is applied when calling `ctx.delay()` without arguments when composing your mocked response. Like so:

```js
rest.get('/user', (req, res, ctx) => res(ctx.json(...))) // no delay

rest.post('/book', (req, res, ctx) => res(ctx.delay(), ctx.json(...))) // random realistic delay

rest.delete('/user', (req, res, ctx) => res.(ctx.delay(500), ctx.json(...))) // exact delay duration (500ms)
```

> We have been noticing a bit of lag for MSW to respond to requests.

Does this happen when running MSW in a browser or in NodeJS? Or in both?  Also, what kind of delay are you encountering (ms/s)?

While there's no default delay set on the mocked response, the request/response lifecycle can still take time:

1. Request needs to happen. Your request client (fetch/etc.) is called and dispatches the request.
2. Service Worker intercepts the request, clones it, and sends it back to the client-side code. 
3. Client-side code looks up a relevant request handler. This may take some time if you have a significant amount of handlers defined.
4. Client-side sends back a found relevant request handler to the worker.

## How to debug this?

### NodeJS

If you're experiencing this in NodeJS, try running your testing script with `DEBUG=*`:

```bash
$ DEBUG=* npm test
```

This prefix will output additional debugging information about what happens internally and how much time certain operations take. 

### Browser

There is no `DEBUG` support in a browser, but you can debug what takes time by looking into the "Network" tab of your browser's DevTools, locating the problematic request, and opening its "Timing" tab. There you can find how much time various request/response operations take. This may hint at what's going wrong.

**Comment by philals** [comment]
Thanks for the detailed response. 

We figured the delay we are noticing was due to the points you outlined.

These tests are React Testing Library running against a largish Create React App. So that would be these are running in NodeJS. Any delay in the browser is not really an issue for us.

The main issue this is causing is in our tests.

We notices a variable delay of up to 200ms in some cases. RTL's default timeout is 1000ms. So if you have 4 cascading requests on a page, then RTL would timeout. 

In some cases we are migrating from mock-fetch which resolves instantly.

Resolutions for us were:

- Look at our tests and make sure we are awaiting things correctly.
- Changing RTL's default wait timeout to 5 seconds

We'll look a little further into it if we start noticing things like this again.

**Comment by philals** [comment]
PS Thanks for making MSW. We love the developer experience of writing tests and being able to use those handlers in browser!

**Reply by kettanaito** [reply]
We are happy to hear such feedback! Thank you, @philals.
